simpleline/event_loop/main_loop.py

```diff
@@ -161,135 +155,96 @@ def _mainloop(self):
 
         Process signals enqueued by the `self.enqueue_signal()` method. Call handlers registered to the signals by
         the `self.register_signal_handler()` method.
 
-        When `return_after` is specified then wait to the point when this signal is processed. This could be after
-        some more signals was processed because of recursion in calls.
-        Without `return_after` parameter this method will return after all queued signals will be processed.
+        When `return_after` is specified then wait to the point when this signal is processed.
+        NO warranty that this method will return immediately after the signal was processed!
+
+        Without `return_after` parameter this method will return after all queued signals with the highest priority
+        will be processed.
 
         The method is NOT thread safe!
 
         :param return_after: Wait on this signal to be processed.
         :type return_after: Class of the signal.
         """
+        super().process_signals(return_after)
+        if return_after is not None:
+            self._process_signals_with_return(return_after)
+        else:
+            self._process_signals_iteration()
+
+    def _process_signals_with_return(self, return_after):
+        """Process signals until the return_after signal was processed.
+
+        Or the loop quited.
+        """
         # get unique ID when waiting for the signal
         unique_id = self._register_wait_on_signal(return_after)
 
-        while self._can_process_signals() or return_after is not None:
+        while self._run_loop:
             signal = self._active_queue.get()
-            # all who is waiting on this signal can stop waiting
-            self._mark_signal_processed(signal)
 
             # do the signal processing (call handlers)
             self._process_signal(signal)
 
+            # all who is waiting on this signal can stop waiting
+            self._mark_signal_processed(signal)
+
             # was our signal processed if yes, return this method
             if self._check_if_signal_processed(return_after, unique_id):
                 return
 
-    def _can_process_signals(self):
-        """To proceed signal processing, these conditions must be true."""
-        return not self._active_queue.empty() and self._run_loop
+    def _process_signals_iteration(self):
+        """Process queued signal and then return."""
+        priority = None
+
+        while not self._active_queue.empty() and self._run_loop:
+            if priority is None:
+                # take first signal to find out the highest priority in queue
+                signal = self._active_queue.get()
+            else:
+                # get signal with this priority only
+                signal = self._active_queue.get_top_event_if_priority(priority)
+
+            # Signal with this priority is not available anymore
+            if signal is None:
+                return
+
+            priority = signal.priority
```

Comment:
`pritority` could be set in the `if priority is None` branch, because it cannot change during the iteration, right?

Reply:
Yeah it could but why? I see it as a pointless line .

Reply:
Fixed.